### ISSUE TITLE
Queries with excluded tags take a long time to process

### ISSUE THREAD

**Issue by FredericaBernkastel** [mserajnik/hydrusrv#4]
On your test server, query like [`scenery -purple_eyes`](https://hydrusrvue.mser.at/#/files?tags%5B0%5D=scenery&tags%5B1%5D=-purple_eyes) takes significantly longer to process.
<br>
And on my test data (20k x 695k),

**Normal timing (is it?):**
`GET /api/files?tags[0]=loli&tags[1]=pussy&sort=id&page=1 200 744.552 ms`

**With excluded tag:**
`GET /api/files?tags[0]=loli&tags[1]=pussy&tags[2]=-penis&sort=id&page=1 200 536324.516 ms`

**Comment by mserajnik** [comment]
Yes, I've noticed this as well. It's particularly bad in this case on my test data because every one of the 2k files has the `scenery` tag. If you try something like `tree -house` it's a lot faster. So it gets slower the more files are included (probably why it takes so long on your test data as well). Simply excluding via `-scenery` seems to be quite quick even though every file, so the combination seems to be what makes it slow.

It thankfully doesn't affect performance of regular queries that don't have excluded tags, so it's not really a regression since excluding tags was a new feature, not one that suddenly got slower (I'll rename the issue to reflect that better).

I'm not sure how it would be possible to make this a lot faster, given the current data structure (but the structure itself is flexible since it's always built from scratch as you know, so we could add/adjust if needed). But if you want to take a look at it, there are two relevant sections, [this function](https://github.com/mserajnik/hydrusrv/blob/3.13.0/server/models/files.js#L133-L180) (which is used when only having exclude tags) and [this part](https://github.com/mserajnik/hydrusrv/blob/3.13.0/server/models/files.js#L73-L88) (which is used when having tags and exclude tags).

**Comment by FredericaBernkastel** [comment]
```
> This SQLite database is used to store a temporary copy of the hydrus server data in an optimized form (for faster on-demand querying).
```
Now this is a complete lie xD

Can you explain why you did not use hydrus's itself data model?
Wasn't able to make pages work?

**Comment by mserajnik** [comment]
It‘s not a lie, just a bit badly phrased. xD

It‘s an „optimized form“ for what hydrusrv needs to be able to do (deliver paginated results while still having the functionality of sorting by multiple namespaces). This wouldn‘t have been possible with the way hydrus structures it‘s data. I found no way to get the sorting by multiple namespaces working that way (especially not with any kind of decent performance). I‘ve tried, if you look back at the `1.x` releases.

The „faster on-demand querying“ isn‘t related to the „stored in an optimized form“, but the fact that it would be way slower if I did the necessary transformations on the fly when the user sends a requests.

hydrus server can‘t do paginated results, it always returns everything for a given query which is obviously not a good fit for a Web-based software.

I tried to make my data structure as lean as possible, but there‘s probably some improvement that can be done (as seen in this issue). :)

**Comment by FredericaBernkastel** [comment]
better-sqlite3 doesn't support backup api, so i cannot dump in-memory database to file, and look at it.
Just by reading the code, it's hard to understand what's happening on the whole.

Can you show ERD diagram of your data model?

**Comment by mserajnik** [comment]
Sure, I don't have one ready, but I'll draw one for you (minus the user-related stuff, as that's not relevant to this).

**Comment by mserajnik** [comment]
<img width="895" alt="screenshot 2018-11-10 at 01 56 15" src="https://user-images.githubusercontent.com/28833155/48295411-d3cfd400-e48b-11e8-9430-05b092d208d4.png">

There you go. Just realized, the new `tags_id` field on `hydrusrv_files` from the latest changes doesn't have an index yet, this will probably negatively affect performance to some degree. I'll correct that in a hotfix.

The `file_count` field on the `hydrusrv_tags` table is populated when filling the temporary tables because that's faster than getting the count via joining with `hydrusrv_files` when the user sends a request (especially since the count is displayed for each tag in the hydrusrvue tags view).

`namespaces` isn't connected to anything because that's really just a table to store all the namespaces that are then served to be able to set the colors in hydrusrvue for now. The actual sorting happens via the `namespace_x` columns directly on the `hydrusrv_files` table.

**Comment by mserajnik** [comment]
I have an idea:

1. Run a query that gets the file IDs via included tags
2. Run a query that gets the file IDs via excluded tags
3. Substract the second set from the first (so only the relevant file IDs remain)
4. Fetch the files via this remaining ID set

I think this should be faster than the subselect? Will test later.

**Comment by mserajnik** [comment]
Actually, I think I found an even simpler solution: https://github.com/mserajnik/hydrusrv/commit/3eaf57b3a56ce57f4837b604e148b2a0f43ba406

Could you give the latest commit on `develop` a try? Currently not at home and can't update my demo server to try it there.

I think the `NOT IN` check should be way faster because if I understand correctly, the current `NOT EXISTS` subselect check needs to run for every potential file ID since it has a reference to it in the via `hydrusrv_mappings.file_tags_id = hydrusrv_files.tags_id`. This new subselect should only need to run once for the whole query (I think?).

**Comment by FredericaBernkastel** [comment]
Nice, normal query now takes 700ms, with excluded tag - 1300ms.
Which is far from perfect, i'll think about better solution.

**Comment by mserajnik** [comment]
Another idea: what do you think about caching the results into a cache table, after a query has been made?

E.g., if a query `GET /api/files?tags[0]=scenery&tags[1]=-purple_eyes&sort=id&page=1` comes in, I generate a hash from that query and check if a row for that hash already exists in the cache table. If yes, get the result directly from the cache table. If no, query normally (and return the result) and save the result into the cache table. The cache table would then be emptied with the next update.

This will make the initial queries marginally slower (I don't think by much), but would probably result in a way faster response if the same query is made again.

Of course, this could also be done for the tags listing queries.

**Comment by FredericaBernkastel** [comment]
Bad idea. If query takes 700ms for 695k mappings, how long it will take for 20m mappings?
And if number of combinations is a factorial from unique tags, 99.9999% of queries won't exist in cache yet, doesn't matter how big your cache is.

I'm not sure if my idea is better of worse, need to write it first in order to test.

We will store 2 mapping tables.
First one with relationship "file-to-tags`,
Second with relationship "tag-to-files".

To get files containing all specified tags, we will use recursion (you will need to build such query dynamically):
```sql
/* recursive mappings reduction proto */
select `file`, `tag`
from hydrusrv_mappings_file_to_tags where `file` in
(select `file` from /* prevous iteration */ where `tag` = ?)
```

**For example, files, having all 1,2,3 tags:**
```sql
select `file`, `tag`
from hydrusrv_mappings_file_to_tags where `file` in
(select `file` from (
    select `file`, `tag`
    from hydrusrv_mappings_file_to_tags where `file` in
    (select `file` from (
        select `file`, `tag`
        from hydrusrv_mappings_file_to_tags where `file` in
        (select `file` from hydrusrv_mappings_file_to_tags 
        where `tag` = 1)
    ) where `tag` = 2)
) where `tag` = 3)
```

To exclude tags, we will use tag-to-files relationship, simply adding to the top level
```sql
and `file` not in
(select `file`
from hydrusrv_mappings_tag_to_files where `tag` in (4,5))
```

Final query will be wrapped with `select distinct`, ordering, and pagination stuff:
```sql
select distinct `file` from (
    select `file`, `tag`
    from hydrusrv_mappings_file_to_tags where `file` in
    (select `file` from (
        select `file`, `tag`
        from hydrusrv_mappings_file_to_tags where `file` in
        (select `file` from (
            select `file`, `tag`
            from hydrusrv_mappings_file_to_tags where `file` in
            (select `file` from hydrusrv_mappings_file_to_tags 
            where `tag` = 1)
        ) where `tag` = 2)
    ) where `tag` = 3)
    
    and `file` not in
    
    (select `file`
    from hydrusrv_mappings_tag_to_files where `tag` in (4,5)) 
)

order by `file` asc

limit 45

offset 0
```

**Comment by mserajnik** [comment]
> If query takes 700ms for 695k mappings, how long it will take for 20m mappings?

Actually, it's pretty quick. I don't have @Robert99943's DB at hand anymore, but I don't remember having any performance issues there. The mappings are already reduced by the provided tags, so you don't check if the file ID exists in one of the 20m mappings, just if it exists in one of the mappings that reference a tag ID.

**Comment by mserajnik** [comment]
@FredericaBernkastel 

I know this is probably still WIP, but just because I saw it in https://github.com/FredericaBernkastel/hydrusrv/commit/fe4f29cd10d5b80673214999a244b129402b3af8:

The way you've rewritten the query in the `getByExcludeTags()` function, it would exclude files that don't have any tags at all (because there are no entries in `hydrusrv_mappings` for such files). Such files without tags should still be in the result when excluding tags (since they obviously don't have those tags).

Otherwise, I'm curious how this will perform, I didn't even know that SQLite has the `EXCEPT` operator. 👍

**Comment by mserajnik** [comment]
Whoops, just noticed you already opened a PR. Still wish GitHub would display that in the dashboard. :/

**Comment by FredericaBernkastel** [comment]
As i mentioned before, the way you implemented query caching is unuseful.
The number of combinations is a stupidly big number, difined as
`tags! * pages * sorting options`

Probability, that one user will do the same query as another user is tends to zero.

In this case, it's better for you to ditch out query caching from hydrusrv, and implement it on the client side.
The only thing you have to worry about, is properly invalidate browser cache, if hydrusrv did update.

**Comment by FredericaBernkastel** [comment]
Got it.

Speaking of usecases, if i'll decide to make a big public gallery, like danbooru, is there a easy way to make registration ***optional***. So, people won't be forced to register in order to view images?

**Comment by mserajnik** [comment]
Not easily.

Server-side it would need a way to disable the authentication middleware from protected routes (except for routes that will still require a logged in user, e.g., changing username/password etc.). This wouldn't be that hard to do, just requires some rewriting of how the middlewares are structured so the authentication checks can be cleanly included/excluded if needed.

The client is a bit more complicated. It would need an option to disable the authentication checks (both the ones that are sent against the server and the client-side ones against the authentication cookie). This isn't as simple as it sounds, as there are quite a few different checks/automatic logout handling in place. Adjusting the views itself (so a non-logged in user would be able to access the settings view, but not the user view or the logout button) would be trivial though.

That said, hydrusrv doesn't collect any data about the user aside from any request logging that might be set up via the config option in hydrusrv itself or in other applications like nginx (or however the setup looks like) and would still be in place regardless of if a user is logged in or not. It doesn't save which requests a user makes or anything of that sort. The username also doesn't have to be an email address or anything that could identify a real person behind it.

So I think you can disregard the privacy question, if that's a concern. But I understand that registration and login is an unnecessary inconvenience if you just want to make it public for everyone to access.

If you absolutely need it, I can tackle that, but you'd need to give me at least a few weeks. I have some other projects in the pipeline that take priority and if I implemented this, I'd want to make sure to make it as clean as possible (and probably also simplify some things regarding authentication handling in the client).

**Comment by mserajnik** [comment]
Closed via #5 (for some reason, GitHub didn't close it automatically via [closing keyword](https://help.github.com/articles/closing-issues-using-keywords/)).
